Add sloppy argument to audiofile.duration()

hagenw opened the pull request:
At the moment the only focus of `audiofile.duraton()` is on samplewise precision of the returned duration, to make sure the duration returned by `audiofile.duration()` and `audiofile.samples()` always match.

For files that need to be decoded first (e.g. MP3), it first decodes them to WAV and then measures the duration.

This has two disadvantages:
* it is slow
* it might return different results on different machines if they use different decoders

To address this I add the `sloppy=False` argument to `audiofile.duration()`. If set to `True` it returns the duration is stored in the header of the audio file and only calculates it if nothing is given in the header.

![image](https://user-images.githubusercontent.com/173624/128182303-faee68c6-f5c6-4a54-8278-222d760369ff.png)

I also updated the benchmark page and included the new results there:

![image](https://user-images.githubusercontent.com/173624/128182419-830d2cc9-7b76-4e3f-9dfe-8f1f0a03a6eb.png)

hagenw commented:
Interestingly, it turns out that duration returned with `sloppy=True` is not independent of the machine.
The test fails as we get a different result for `gs-16b-1c-44100hz.opus` on macOS than on Ubuntu.

Maybe `sox.file_info.duration()` does not read from the header.

hagenw commented:
I makred it as WIP and will change back, when I have fixed the error.

hagenw commented:
> Interestingly, it turns out that duration returned with `sloppy=True` is not independent of the machine.
> The test fails as we get a different result for `gs-16b-1c-44100hz.opus` on macOS than on Ubuntu.
> 
> Maybe `sox.file_info.duration()` does not read from the header.

For me `soxi` does not support the OPUS file, which means the different results are due to `mediainfo`.
There is not much we can do about.

One solution would be to change the docstring and state that the durations might also differ between machines
and/or round the duration to have only 1-2 digits after the dot.

phtephanx commented:
> One solution would be to change the docstring and state that the durations might also differ between machines
> and/or round the duration to have only 1-2 digits after the dot.

I don't have a strong opinion, both solutions look good to me.

It might be interesting, though, to have, at least, one configuration where results are machine-independent. Are we sure that rounding to 1-2 decimals is sufficient to be machine-independent? It sounds large enough. If not, your first proposal might be better.

hagenw commented:
> > One solution would be to change the docstring and state that the durations might also differ between machines
> > and/or round the duration to have only 1-2 digits after the dot.
> 
> I don't have a strong opinion, both solutions look good to me.
> 
> It might be interesting, though, to have, at least, one configuration where results are machine-independent. Are we sure that rounding to 1-2 decimals is sufficient to be machine-independent? It sounds large enough. If not, your first proposal might be better.

I was digging a little bit deeper and the reason why it is not machine-independent at the moment is the following: `soxi` can be compiled with or without MP3 support. As it is the fastest way to access the metadata we first try to use it to read the duration of the file. If it is not available we use `mediainfo`, which we use for most exotic formats like MP4. Interestingly, both commands return different duration values (sic!). So, I don't know why, and I'm not very motivated to find out why, but this means the duration depends on what programs you have installed on your machine.

One way around this would be to use only `mediainfo` and never `soxi`. The disadvantage is that it is slower, and for MP3 files it is even slower then first converting to WAV and counting samples:

![benchmark_mp3-mp4_info](https://user-images.githubusercontent.com/173624/128155327-67018edb-afe7-46d9-99bb-f1aa66e52d16.png)

The other problem is that for some file formats you don't have to store the duration in the header. In this case we need to calculate the duration the way we did before, which would also break machine-independence.

So, what we have to decide here is what should be our highest priority:
1. speed
2. machine independence

If we go with 1. I would stay with first trying to use `soxi` (which is a factor 2 faster for MP3 files) and just adjust the tests that they check only for one digit precision and mention in the docs that the duration might depend on your system.
If we go with 2. I would only use `mediainfo` to get duration for non SND formats and state in the docs that it is machine independent as long as the duration is stored in the header (I do it already, by indicating that it falls back to `sloppy=False` in this case).

Any preference?

phtephanx commented:
> If we go with 1. I would stay with first trying to use `soxi` (which is a factor 2 faster for MP3 files) and just adjust the tests that they check only for one digit precision and mention in the docs that the duration might depend on your system.
> If we go with 2. I would only use `mediainfo` to get duration for non SND formats and state in the docs that it is machine independent as long as the duration is stored in the header (I do it already, by indicating that it falls back to `sloppy=False` in this case).

Thanks for the details! I'd prefer **speed**. The time difference of 0.1 [s] between sloppy and non-sloppy for `mp4` is noticeable for lots of files. Machine independence could be reached by the user subsequently by rounding to 1 decimal. Apart from this, the different result occurred on macOS, which not used often in research, whereas the different ubuntu versions gave the same result.

hagenw commented:
I agree, I updated now the code with speed as first priority. I also updated the benchmark page in the docs, see the description of this pull request.

phtephanx commented:
> I agree, I updated now the code with speed as first priority. I also updated the benchmark page in the docs, see the description of this pull request.

I reviewed docs, logic of function and tests. Please merge.
